### PR TITLE
Support BloomFilterMightContain expression

### PR DESCRIPTION
**Pull request by jlowe**
Contributes to #7803.  Depends on NVIDIA/spark-rapids-jni#1269.

This adds support for BloomFilterMightContain, which is the filter side needed for bloom-filter accelerated joins.  The Bloom filter arrives as a scalar BINARY from a subquery, which is serialized as a big-endian buffer due to the use of DataInputStream.  The GPU kernel operates on this big-endian format directly.

### CODE REVIEW

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+          withResource(s.getBase.getListAsColumnView) { childView =>
+            require(childView.getType == DType.UINT8, s"expected UINT8 got ${childView.getType}")
+            deserialize(childView.getData)
+          }
+        case t => throw new IllegalArgumentException(s"Expected binary scalar, found $t")
+      }
+    }
+  }
+
+  def deserialize(data: BaseDeviceMemoryBuffer): GpuBloomFilter = {
```
Comment (revans2) [inline sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala:91]: We don't support serialize yet, but is it ever possible for data to be null? Even for a null scalar? What about if the scalar was null would there still be a header for it? Like what happens if none of the input to the join made it through the first filter stage (checking for nulls) what might happen then?

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    if (s == null) {
+      null
+    } else {
+      s.dataType match {
+        case BinaryType =>
+          withResource(s.getBase.getListAsColumnView) { childView =>
+            require(childView.getType == DType.UINT8, s"expected UINT8 got ${childView.getType}")
+            deserialize(childView.getData)
+          }
+        case t => throw new IllegalArgumentException(s"Expected binary scalar, found $t")
```
Comment (revans2) [inline sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala:86]: So in the check bloom filter code we say that we support NullType, but if it shows up here we will throw an exception.

File: tests/src/test/spark330/scala/com/nvidia/spark/rapids/BloomFilterAggregateQuerySuite.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+               |FROM $table
+              """.stripMargin
+          df.createOrReplaceTempView(table)
+          try {
+            installSqlFuncs(df.sparkSession)
+            df.sparkSession.sql(sqlString)
+          } finally {
+            uninstallSqlFuncs(df.sparkSession)
+          }
+        })
```
Comment (revans2) [inline tests/src/test/spark330/scala/com/nvidia/spark/rapids/BloomFilterAggregateQuerySuite.scala:99]: I don't see any tests for nulls.  Null input values or null bloom filter. I know that these are corner cases that should not exist in our real world use cases, but I would like to make sure that we are covered because the CPU thinks that they need to support these cases.

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala:91]: If the scalar from the subquery is null then we won't try to create a GpuBloomFilter at all.  That gets cutoff early in GpuBloomFilter.apply which returns null if the scalar to construct the filter is null.  GpuBloomFilterMightContains handles a null bloomFilter, returning nulls if that occurs.

Comment (revans2) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala:91]: if `Scalar.isValid` returns `false`. Not if the value returned for the scalar is null. The latter should never happen.

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+   * Given an input column of longs, return a boolean column with the same row count where each
+   * output row indicates whether the corresponding input row may have been placed into this
+   * Bloom filter. A false value indicates definitively that the value was not placed in the filter.
+   */
+  def mightContainLong(col: ColumnVector): ColumnVector = {
+    require(col.getType == DType.INT64, s"expected longs, got ${col.getType}")
+    withResource(spillableBuffer.getDeviceBuffer()) { buffer =>
+      buffer.incRefCount()
+      withResource(new BloomFilter(numHashes, numFilterBits, buffer)) { filter =>
+        filter.probe(col)
```
[the remark this answers is not in the document]
Comment (jlowe) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala]: This is trivial, just creating a wrapper object to call a method on it.  There's no computation involved in the constructor here.  In the most recent version I moved to a one-shot static method that was recently exposed so we don't need to do the somewhat silly object creation just for the one method call.  That leaves us with just the spillable bloom filter device buffer which is only locked down into memory for the duration of this probe call.

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala:91]: Ah yes, followed the Spark code way too closely on this.  The null handling should be fixed with the latest commit.

File: tests/src/test/spark330/scala/com/nvidia/spark/rapids/BloomFilterAggregateQuerySuite.scala (hunk 3 above)
Comment (jlowe) [reply tests/src/test/spark330/scala/com/nvidia/spark/rapids/BloomFilterAggregateQuerySuite.scala:101]: Null tests were added but are not passing currently because NVIDIA/spark-rapids-jni#1269 currently doesn't support null rows (actually throws an exception if they are seen).  It was built with the assumption input would always be from xxhash64 which cannot produce nulls.  However I agree it's prudent to handle nulls just in case this expression gets used outside of an xxhash64 context.  @nvdbaranec is currently updating the spark-rapids-jni  PR to handle nulls.

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilterMightContain.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        if (value == null || value.dataType == NullType) {
+          GpuColumnVector.fromNull(batch.numRows(), dataType)
+        } else {
+          GpuColumnVector.from(bloomFilter.mightContainLong(value.getBase), BooleanType)
+        }
+      }
+    }
+  }
+
+  override def close(): Unit = {
```
Comment (revans2) [inline sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilterMightContain.scala:99]: When is this called?

Comment (jlowe) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilterMightContain.scala:99]: When `bloomFilter` is lazily evaluated here: https://github.com/NVIDIA/spark-rapids/pull/8775/files#diff-95125ce9b0fef86d71d7b20aa1de5d59ab49239f26d13536961e436a6d74e44cR47

**Review by revans2 — APPROVED** [review]

**Review by revans2 — DISMISSED** [review]

**Comment by jlowe** [comment]
Updated to new BloomFilter interface in NVIDIA/spark-rapids-jni#1303.

**Comment by jlowe** [comment]
build

**Review by revans2 — APPROVED** [review]
